### PR TITLE
add support for IIJ PTA(Period of Time Authentication) module token.

### PR DESCRIPTION
**Pull request by kuroishi**
Hi, 

We develop a secure token using HTTP query parameter that is called IIJ PTA, which is released at https://github.com/iij/pta as free software.

We think that your nginx-secure-token-module is very helpful for video streaming such as HLS, MPEG-DASH and we would like to use it with our PTA module. I've implemented it so I would appreciate if you merge this patch.

Regards,

### CODE REVIEW

**Comment by kaltura-hooks** [comment]
Hi @kuroishi,
Thank you for contributing this pull request!
Please sign the Kaltura CLA so we can review and merge your contribution.
Learn more at http://bit.ly/KalturaContrib

File: iij/ngx_http_secure_token_iijpta.c [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+	/* sizeof("pta=") returns 5, includes null termination. */
+	qparam = ngx_pnalloc(r->pool, sizeof("pta=") + ((out_len1 + out_len2) * 2));
+	if (qparam == NULL)
+	{
+		return NGX_ERROR;
+	}
+
+	v->data = qparam;
+	qparam = ngx_copy(qparam, "pta=", sizeof("pta=") - 1);
+	qparam = ngx_hex_dump(qparam, out, out_len1 + out_len2);
```
Comment (erankor) [inline iij/ngx_http_secure_token_iijpta.c]: the format of the token is to hex encode everything? sounds like a big waste... why doesn't it use base64 or something?

File: iij/ngx_http_secure_token_iijpta.c [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	
+	deadline = now + token->timelimit;
+	deadline = htobe64(deadline);
+	memcpy(&in[CRC32_SIZE], &deadline, sizeof(deadline));
+	memcpy(&in[CRC32_SIZE + DEADLINE_SIZE], token->path.data, token->path.len);
+	crc = ngx_crc32_long(&in[CRC32_SIZE], DEADLINE_SIZE + token->path.len);
+	crc = htobe32(crc);
+	memcpy(in, &crc, sizeof(crc));
+
+	EVP_EncryptInit(&ctx, EVP_aes_128_cbc(), token->key.data, token->iv.data);
```
Comment (erankor) [inline iij/ngx_http_secure_token_iijpta.c]: missing checks for errors in these 3 calls

File: iij/ngx_http_secure_token_iijpta.c [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	EVP_EncryptUpdate(&ctx, out, &out_len1, in, in_len);
+	EVP_EncryptFinal(&ctx, out + out_len1, &out_len2);
```
Comment (erankor) [inline iij/ngx_http_secure_token_iijpta.c]: I think a nicer pattern (instead of the 2 length vars) is to use a pointer that advances by the output length, see ngx_http_secure_token_crypt

File: iij/ngx_http_secure_token_iijpta.c [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+	// validate required params
+	if (token->key.data == NULL)
+	{
+		ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+			"\"key\" is mandatory for iijpta tokens");
+		return NGX_CONF_ERROR;
+	}
+
+        if (token->key.len != 16)
```
Comment (erankor) [inline iij/ngx_http_secure_token_iijpta.c]: once bounds are added (as commented above) these checks can be removed

File: iij/ngx_http_secure_token_iijpta.c [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	{ ngx_string("iv"),
+	NGX_CONF_TAKE1,
+	ngx_http_secure_token_conf_set_hex_str_slot,
+	0,
+	offsetof(ngx_secure_token_iijpta_token_t, iv),
+	NULL },
+
+	{ ngx_string("path"),
+	NGX_CONF_TAKE1,
+	ngx_conf_set_str_slot,
```
Comment (erankor) [inline iij/ngx_http_secure_token_iijpta.c]: better change to complex value, in order to enable the user to use nginx variables when defining it. see the acl parameter in Akamai. would also be a good idea to rename it to acl, for consistency (it has exactly the same meaning...)

File: iij/ngx_http_secure_token_iijpta.c [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+#include "../ngx_http_secure_token_filter_module.h"
+#include "../ngx_http_secure_token_utils.h"
+
+#include <time.h>
+#include <openssl/hmac.h>
+#include <openssl/evp.h>
+
+// constants
+#define CRC32_SIZE    4
+#define DEADLINE_SIZE 8
```
Comment (erankor) [inline iij/ngx_http_secure_token_iijpta.c]: instead of CRC32_SIZE & DEADLINE_SIZE, it will probably make nicer code to define a struct - 
```
typedef struct {
  u_char size[4];
  u_char expiry[8];
} ngx_http_secure_token_iijpta_header_t;
```
(I also prefer the name `expiry` over `deadline`)

File: iij/ngx_http_secure_token_iijpta.c [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+	ngx_http_request_t *r,
+	ngx_http_variable_value_t *v,
+	uintptr_t data)
+{
+	EVP_CIPHER_CTX ctx;
+	uint32_t crc;
+	ngx_secure_token_iijpta_token_t* token = (void*)data;
+	u_char* in;
+	size_t in_len  = CRC32_SIZE + DEADLINE_SIZE + token->path.len;
+	u_char *qparam;
```
Comment (erankor) [inline iij/ngx_http_secure_token_iijpta.c]: usually in nginx, such moving pointers are named `p`

**Comment by kuroishi** [comment]
thank you for your feedback. I'll fix them.

File: iij/ngx_http_secure_token_iijpta.c [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+} ngx_secure_token_iijpta_token_t;
+
+// globals
+static ngx_command_t ngx_http_secure_token_iijpta_cmds[] = {
+	{ ngx_string("key"),
+	NGX_CONF_TAKE1,
+	ngx_http_secure_token_conf_set_hex_str_slot,
+	0,
+	offsetof(ngx_secure_token_iijpta_token_t, key),
+	NULL },
```
[the remark this answers is not in the document]
Comment (kuroishi) [reply iij/ngx_http_secure_token_iijpta.c]: Hey, Where is the post_handler in ngx_http_secure_token_encrypt_uri_key_bounds called? Doesn't the ngx_http_secure_token_conf_set_hex_str_slot have the post_handler call does it?
The logic of string length validation seems not to work I think.

I would like your comment.

[the remark this answers is not in the document]
Comment (erankor) [reply iij/ngx_http_secure_token_iijpta.c]: You're right, fixed here #70

File: config [hunk 9]
```diff
@@ -8,6 +8,8 @@ SEC_TOKEN_DEPS="$ngx_addon_dir/akamai/ngx_http_secure_token_akamai.h
 				$ngx_addon_dir/cht/ngx_http_secure_token_cht_commands.h               \
 				$ngx_addon_dir/cloudfront/ngx_http_secure_token_cloudfront.h          \
 				$ngx_addon_dir/cloudfront/ngx_http_secure_token_cloudfront_commands.h \
+				$ngx_addon_dir/iij/ngx_http_secure_token_iijpta.h                     \
+				$ngx_addon_dir/iij/ngx_http_secure_token_iijpta_commands.h                     \
```
[the remark this answers is not in the document]
Comment (kuroishi) [reply config]: fix at 93b19fb

File: iij/ngx_http_secure_token_iijpta.c [hunk 10]
```diff
@@ -0,0 +1,191 @@
+#include "ngx_http_secure_token_iijpta.h"
+#include "../ngx_http_secure_token_filter_module.h"
+#include "../ngx_http_secure_token_utils.h"
+
+#include <time.h>
+#include <openssl/hmac.h>
```
[the remark this answers is not in the document]
Comment (kuroishi) [reply iij/ngx_http_secure_token_iijpta.c]: fix at afc729c

File: config [hunk 11]
```diff
@@ -8,6 +8,8 @@ SEC_TOKEN_DEPS="$ngx_addon_dir/akamai/ngx_http_secure_token_akamai.h
 				$ngx_addon_dir/cht/ngx_http_secure_token_cht_commands.h               \
 				$ngx_addon_dir/cloudfront/ngx_http_secure_token_cloudfront.h          \
 				$ngx_addon_dir/cloudfront/ngx_http_secure_token_cloudfront_commands.h \
+				$ngx_addon_dir/iij/ngx_http_secure_token_iijpta.h                     \
```
[the remark this answers is not in the document]
Comment (kuroishi) [reply config]: fix at a15ca5e

File: iij/ngx_http_secure_token_iijpta.c [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+	NULL },
+};
+
+static ngx_int_t
+ngx_secure_token_iijpta_get_var(
+	ngx_http_request_t *r,
+	ngx_http_variable_value_t *v,
+	uintptr_t data)
+{
+	EVP_CIPHER_CTX ctx;
```
[the remark this answers is not in the document]
Comment (kuroishi) [reply iij/ngx_http_secure_token_iijpta.c]: fix at 877c4da

File: iij/ngx_http_secure_token_iijpta.c [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+	time_t now;
+	uint64_t deadline;
+
+	in = ngx_pnalloc(r->pool, in_len);
+	if (in == NULL)
+	{
+		return NGX_ERROR;
+	}
+
+	now = time(NULL);
```
[the remark this answers is not in the document]
Comment (kuroishi) [reply iij/ngx_http_secure_token_iijpta.c]: fix at 4a3dc9e

File: iij/ngx_http_secure_token_iijpta.c (hunk 3 above)
Comment (kuroishi) [reply iij/ngx_http_secure_token_iijpta.c]: fix at b7e8da9

File: iij/ngx_http_secure_token_iijpta.c (hunk 4 above)
Comment (kuroishi) [reply iij/ngx_http_secure_token_iijpta.c]: fix at 066424c

File: iij/ngx_http_secure_token_iijpta.c (hunk 6 above)
Comment (kuroishi) [reply iij/ngx_http_secure_token_iijpta.c]: fix at a27d59a

File: iij/ngx_http_secure_token_iijpta.c (hunk 7 above)
Comment (kuroishi) [reply iij/ngx_http_secure_token_iijpta.c]: fix at 582635f

**Comment by kuroishi** [comment]
Sorry for replying so late.
Please check the new PR.
Thanks,

File: iijpta/ngx_http_secure_token_iijpta.c [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+	{ ngx_string("end"),
+	NGX_CONF_TAKE1,
+        ngx_http_secure_token_conf_set_time_slot,
+	0,
+	offsetof(ngx_secure_token_iijpta_token_t, end),
+	NULL },
+};
+
+static char *
+ngx_conf_check_byte_len_bounds(ngx_conf_t *cf, void *post, void *data)
```
Comment (erankor) [inline iijpta/ngx_http_secure_token_iijpta.c]: I prefer to move the existing ngx_conf_check_str_len_bounds function from filter_module.c to utils.c and use it here instead of duplicating

File: iijpta/ngx_http_secure_token_iijpta.c [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+	&ngx_http_secure_token_iijpta_key_bounds },
+
+	{ ngx_string("iv"),
+	NGX_CONF_TAKE1,
+	ngx_http_secure_token_conf_set_hex_str_slot,
+	0,
+	offsetof(ngx_secure_token_iijpta_token_t, iv),
+	&ngx_http_secure_token_iijpta_iv_bounds },
+
+	{ ngx_string("path"),
```
Comment (erankor) [inline iijpta/ngx_http_secure_token_iijpta.c]: Better rename to acl for consistency with other tokens + change to 'complex value' in order to support variables, such as $uri (see Akamai as an example)

File: iijpta/ngx_http_secure_token_iijpta.c [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+{
+	EVP_CIPHER_CTX *ctx;
+	uint32_t crc;
+	ngx_secure_token_iijpta_token_t* token = (void*)data;
+	u_char* in;
+	ngx_http_secure_token_iijpta_header_t *hdr;
+	size_t in_len  = sizeof(ngx_http_secure_token_iijpta_header_t) + token->path.len;
+	u_char *p;
+	// in_len rounded up to block + one block for padding
+	uint8_t out[((in_len + (16 - 1)) / 16) + 16];
```
Comment (erankor) [inline iijpta/ngx_http_secure_token_iijpta.c]: Some compilers may not accept this (dynamically sized array) better allocate with ngx_pnalloc

File: iijpta/ngx_http_secure_token_iijpta.c [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+	u_char *outp;
+	int out_len;
+	uint64_t end;
+
+	in = ngx_pnalloc(r->pool, in_len);
```
Comment (erankor) [inline iijpta/ngx_http_secure_token_iijpta.c]: `in` is not needed, you can EVP_EncryptUpdate twice - once for header + once for the path. For the crc, you can use ngx_crc32_update on the second call.

File: iijpta/ngx_http_secure_token_iijpta.c [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+	memcpy(hdr->crc, &crc, sizeof(crc));
+
+	ctx = EVP_CIPHER_CTX_new();
+	if (ctx == NULL) {
+	    ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
+			  "ngx_secure_token_iijpta_get_var: EVP_CIPHER_CTX_new failed");
+	    return NGX_ERROR;
+	}
+
+	if (!EVP_EncryptInit(ctx, EVP_aes_128_cbc(), token->key.data, token->iv.data)) {
```
Comment (erankor) [inline iijpta/ngx_http_secure_token_iijpta.c]: Inconsistent { convention, other than code copied from nginx, { should be after newline

File: iijpta/ngx_http_secure_token_iijpta.c [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+	    ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
+			  "ngx_secure_token_iijpta_get_var: EVP_EncryptInit failed");
+	    EVP_CIPHER_CTX_free(ctx);
```
Comment (erankor) [inline iijpta/ngx_http_secure_token_iijpta.c]: better add an 'error' label + use goto, instead of duplicating these 2 lines

File: iijpta/ngx_http_secure_token_iijpta.c [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+
+	in = ngx_pnalloc(r->pool, in_len);
+	if (in == NULL)
+	{
+		return NGX_ERROR;
+	}
+
+	hdr = (ngx_http_secure_token_iijpta_header_t *)in;
+	memcpy(&in[sizeof(hdr)], token->path.data, token->path.len);
+	end = htobe64(ngx_time() + token->end.val);
```
Comment (erankor) [inline iijpta/ngx_http_secure_token_iijpta.c]: end may not be relative, need to do if (token->end.type == NGX_HTTP_SECURE_TOKEN_TIME_RELATIVE) end_time += ngx_time(); as in other tokens

File: ngx_http_secure_token_m3u8.c [hunk 21]
```diff
@@ -1,9 +1,14 @@
 #include "ngx_http_secure_token_encrypt_uri.h"
 #include "ngx_http_secure_token_m3u8.h"
 
-static u_char encryption_key_tag[] = "EXT-X-KEY";
-static u_char ext_media_tag[] = "EXT-X-MEDIA";
-static u_char ext_i_frame_tag[] = "EXT-X-I-FRAME-STREAM-INF";
+static ngx_str_t uri_tags[] = {
```
Comment (erankor) [inline ngx_http_secure_token_m3u8.c]: there are some unrelated commits below this point, better open a new branch/pull so that only your changes will be there.

**Comment by kuroishi** [comment]
I've offered another PR for review to be clear so I close this.
